Added fix for issue https://github.com/ZOSOpenTools/utils/issues/56

HarithaIBM opened the pull request:
Implemented the changes for importenv to handle the sourcing of the dependency port

HarithaIBM committed:
Added fix for issue https://github.com/ZOSOpenTools/utils/issues/56

IgorTodorovskiIBM commented on bin/zopen-importenvs:
```diff
@@ -1,18 +1,87 @@
+depsPath="$HOME/zopen/prod|$HOME/zopen/boot"
+
+# if the number of arguments passed to script is 0,
+# then , the dirs are fetched from zopen/prod and zopen/boot directory
```
```suggestion
# then the ports are fetched from $HOME/zopen/prod and $HOME/zopen/boot directories.
```

HarithaIBM replied:
I have modified it, thank you!

IgorTodorovskiIBM commented on bin/zopen-importenvs:
```diff
@@ -1,18 +1,87 @@
-  . ./.env
-  cd -
-done
+#below code fetches the ~/code/utils/bin path where the common.inc file is under its lib dir
```
Suggest adding a space between # and 'below' for comment consistency

HarithaIBM replied:
I have modified this!

IgorTodorovskiIBM commented on bin/zopen-importenvs:
```diff
@@ -1,18 +1,87 @@
+# then its env files are fetched and sourced, if they are not already sourced.
+if [ $# -eq 0 ]
+then
+     echo "buildenv path is not provided"
```
I suggest changing the echo to "Source projects from <expand depsPath>"

HarithaIBM replied:
I have modified this!

IgorTodorovskiIBM commented on bin/zopen-importenvs:
```diff
@@ -1,18 +1,87 @@
-#!/bin/sh
-#
+#!/bin/bash
```
We may want to change this to /bin/env bash in case bash is not found in /bin since I don't believe it comes with the OS by default.

Also, add a comment that bash is required because of BASH_SOURCE - to locate the path of the sourced shell script

MikeFultonDev replied:
+1 - we should always use 'sh' not 'bash'

HarithaIBM replied:
i have addressed this comment.

IgorTodorovskiIBM commented on bin/zopen-importenvs:
```diff
@@ -1,18 +1,87 @@
-# . importenvs
+#. ./zopen-importenvs <path to buildenv to fetch dependency>
 
-if [ ! -d "${HOME}/zopen/boot/" ]; then
```
I was just thinking that this script could be misused - I can see users trying to calling it like ./zopen-importenvs instead of sourcing it.

We should have a way to detect that if it's not being sourced, print out an error message.

You can use the code from here: https://stackoverflow.com/questions/2683279/how-to-detect-if-a-script-is-being-sourced

MikeFultonDev replied:
+1

HarithaIBM replied:
i have addressed this!

IgorTodorovskiIBM commented on bin/zopen-importenvs:
```diff
@@ -1,18 +1,87 @@
+    #If path is provided then builedenv file is fetched from the provided path 
+    pathForBuild=$1
+    if [ ! -d "${HOME}/zopen/boot/" ] && [ ! -d "${HOME}/zopen/prod/" ]; then
+       echo "boot & prod directories does not exist.  Existing"
```
```suggestion
       echo "boot & prod directories do not exist.  Exiting"
```

HarithaIBM replied:
I have modified this!

IgorTodorovskiIBM commented on bin/zopen-importenvs:
```diff
@@ -1,18 +1,87 @@
+
+    if [ -f ${pathForBuild}/buildenv ]
+    then
+       ZOPEN_GIT_DEPS=$(. ${pathForBuild}/buildenv; echo $ZOPEN_GIT_DEPS); 
```
If the parent environment has ZOPEN_GIT_DEPS set for some reason, will it affect this? If so, we should probably unset it above

HarithaIBM replied:
I have modified this, i have used new variable names to cause no impact.

IgorTodorovskiIBM commented:
Thanks for the PR Haritha!

MikeFultonDev reviewed (changes requested):
see the two +1's on Igor's comments.

HarithaIBM committed:
Addressed comments

HarithaIBM commented:
> Thanks for the PR Haritha!

Thank you for the review, Igor!

HarithaIBM commented:
> see the two +1's on Igor's comments.

I have addressed the comments, please help me understand if any other change is needed. Thank you for the review, Mike!

IgorTodorovskiIBM commented on bin/zopen-importenvs:
```diff
@@ -1,18 +1,96 @@
+if [[ $sourced -ne 1 ]]
+then
+    echo "This script is not sourced."
+    echo "USAGE: . ./zopen-importenvs <path to buildenv to fetch dependency>"
```
My mistake, we should use [] to denote optional paramteer: https://stackoverflow.com/questions/21503865/how-to-denote-that-a-command-line-argument-is-optional-when-printing-usage

HarithaIBM replied:
I have modified it.

IgorTodorovskiIBM commented on bin/zopen-importenvs:
```diff
@@ -1,18 +1,96 @@
+then
+    echo "This script is not sourced."
+    echo "USAGE: . ./zopen-importenvs <path to buildenv to fetch dependency>"
+    exit 6     
```
minor: exit 6 seems a bit random to me.  

future consideration:  We should probably come up with a consistent set of error codes (maybe define in common.inc)

HarithaIBM replied:
I have modified it.

HarithaIBM replied:
Also, i have added the exit 1, condition in another error case too. As of now i have used exit 1 in 2 use cases in this script.

IgorTodorovskiIBM commented on bin/zopen-importenvs:
```diff
@@ -1,18 +1,96 @@
-done
+# below code fetches the ~/code/utils/bin path where the common.inc file is under its lib dir
+export bindir="$( cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd -P )"
+. "${bindir}/lib/common.inc"
```
I wonder if we still need common.inc and if not, then perhaps we can make it a /bin/sh script rather than a bash script.  The reason we try to go with /bin/sh is because not all systems on z/OS have bash, but they all come with sh.

HarithaIBM replied:
Since common.inc is not needed , i have reverted back to /bin/sh

IgorTodorovskiIBM commented on bin/zopen-importenvs:
```diff
@@ -1,18 +1,96 @@
+    #If path is provided then builedenv file is fetched from the provided path 
+    pathForBuild=$1
+    if [ ! -d "${HOME}/zopen/boot/" ] && [ ! -d "${HOME}/zopen/prod/" ]; then
+       echo "boot & prod directories do not exist.  Exiting"
```
I suggest adding the full paths here

HarithaIBM replied:
Ok, i have added full path in echo output.

IgorTodorovskiIBM commented on bin/zopen-importenvs:
```diff
@@ -1,18 +1,96 @@
+
+    if [ -f ${pathForBuild}/buildenv ]
+    then
+       GIT_DEPS=$(. ${pathForBuild}/buildenv; echo $ZOPEN_GIT_DEPS); 
```
I think ZOPEN_GIT_DEPS can still be inherited from the parent if it's not overriden/set in buildenv. I suggest adding
unset ZOPEN_GIT_DEPS
unset ZOPEN_TARBALL_DEPS
unset ZOPEN_TYPE
above this line

HarithaIBM replied:
Sure, i have modified it.

HarithaIBM replied:
Thank you for the review, Igor!

HarithaIBM committed:
Addressed comments

HarithaIBM committed:
Added details to readme file

HarithaIBM commented:
I have added the readme changes as well. Please review. Thank you!

IgorTodorovskiIBM commented on README.md:
```diff
@@ -90,3 +90,14 @@ Run `zopen build` from the root directory of the git repo you would like to buil
+   The path to buildenv is optional.
+
+   If path is provided: the dependencies from the buildenv file are read and the env is sourced from prod/boot directory.
+   Else if path is not provided then env from each of the directory in prod & boot directories are sourced.
```
```suggestion
   Else if path is not provided then the .env from each project directory in the ~/zopen/prod & ~/zopen/boot directories are sourced.
```

HarithaIBM replied:
Yes i have modified, as suggested. Thank you!

IgorTodorovskiIBM reviewed (approved):
LGTM, minor comment

HarithaIBM committed:
Mopdified comments
